### PR TITLE
Add airgap mode

### PR DESCRIPTION
**Pull request by tomgeorge**
This PR addresses https://github.com/eclipse/che/issues/14734

- Add 3 CR fields, `airGapMode`, `airGapContainerRegistryHostname`, `airGapContainerRegistryRepository`
- If `airGapMode` is true, all of the images will be pulled from their airgapped repository location instead of `registry.redhat.io`, `docker.io`, etc.

We are thinking that this doesn't need a docs PR as this will mostly be covered in https://github.com/eclipse/che/issues/13886

### CODE REVIEW

File: pkg/apis/org/v1/che_types.go [hunk 1]
```diff
@@ -20,16 +20,18 @@ import (
...
+// +k8s:openapi-gen=true
 type CheClusterSpec struct {
 	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
 	// Important: Run "operator-sdk generate k8s" to regenerate code after modifying this file
 	Server   CheClusterSpecServer  `json:"server"`
 	Database CheClusterSpecDB      `json:"database"`
 	Auth     CheClusterSpecAuth    `json:"auth"`
 	Storage  CheClusterSpecStorage `json:"storage"`
-	K8SOnly  CheClusterSpecK8SOnly `json:"k8s"`
+	K8s      CheClusterSpecK8SOnly `json:"k8s"`
```
Comment (davidfestal) [inline pkg/apis/org/v1/che_types.go]: Don't you need to change something in the rest of the code ta adapt to this change ?

BTW do you think it would be possible to a distinct PR for openapi support ?

Comment (tomgeorge) [reply pkg/apis/org/v1/che_types.go]: oh whoops, I will change that back

File: pkg/deploy/che_configmap.go [hunk 2]
```diff
@@ -65,6 +65,11 @@ type CheConfigMap struct {
 	PluginRegistryUrl                    string `json:"CHE_WORKSPACE_PLUGIN__REGISTRY__URL,omitempty"`
 	DevfileRegistryUrl                   string `json:"CHE_WORKSPACE_DEVFILE__REGISTRY__URL,omitempty"`
 	WebSocketEndpointMinor               string `json:"CHE_WEBSOCKET_ENDPOINT__MINOR"`
+	CheWorkspacePluginBrokerInitImage    string `json:"CHE_WORKSPACE_PLUGIN__BROKER_INIT_IMAGE,omitempty"`
+	CheWorkspacePluginBrokerUnifiedImage string `json:"CHE_WORKSPACE_PLUGIN__BROKER_UNIFIED_IMAGE,omitempty"`
+	CheServerSecureExposerJwtProxyImage  string `json:"CHE_SERVER_SECURE__EXPOSER_JWTPROXY_IMAGE,omitempty"`
+	CheWorkspaceSidecarImagePullPolicy   string `json:"CHE_WORKSPACE_SIDECAR_IMAGE__PULL__POLICY,omitempty"`
```
Comment (davidfestal) [inline pkg/deploy/che_configmap.go]: > CheWorkspaceSidecarImagePullPolicy
> CheDockerAlwaysPullImage

@l0rd @nickboldt Do you know if Airgap should override these properties to force a given value ?

File: pkg/deploy/che_configmap.go [hunk 3]
```diff
@@ -154,6 +159,8 @@ func GetConfigMapData(cr *orgv1.CheCluster) (cheEnv map[string]string) {
 		CheWebSocketEndpoint:                 wsprotocol + "://" + cheHost + "/api/websocket",
 		WebSocketEndpointMinor:               wsprotocol + "://" + cheHost + "/api/websocket-minor",
 		CheDebugServer:                       cheDebug,
+		CheDockerAlwaysPullImage:             cheDockerAlwaysPullImage,
+		CheWorkspaceSidecarImagePullPolicy:   cheWorkspaceSidecarImagePullPolicy,
```
Comment (davidfestal) [inline pkg/deploy/che_configmap.go]: > CheWorkspaceSidecarImagePullPolicy
> CheDockerAlwaysPullImage

IMO we shouldn't override it by default with a constant, only when airgap is enabled, and possibly with a different value than the default one provided by `che.properties`

I'm not even sure that we have to override them for airgap

File: pkg/deploy/che_configmap.go [hunk 4]
```diff
@@ -225,3 +235,14 @@ func NewCheConfigMap(cr *orgv1.CheCluster, cheEnv map[string]string) *corev1.Con
 		Data: cheEnv,
 	}
 }
+
+func extraImagesConfig(cr *orgv1.CheCluster) map[string]string {
+	extraImages := map[string]string{
+		"CHE_WORKSPACE_PLUGIN__BROKER_INIT_IMAGE":    patchDefaultImageName(cr, cheWorkspacePluginBrokerInitImage),
+		"CHE_WORKSPACE_PLUGIN__BROKER_UNIFIED_IMAGE": patchDefaultImageName(cr, cheWorkspacePluginBrokerUnifiedImage),
+		"CHE_SERVER_SECURE__EXPOSER_JWTPROXY_IMAGE":  patchDefaultImageName(cr, cheServerSecureExposerJwtProxyImage),
+		"CHE_WORKSPACE_SIDECAR_IMAGE__PULL__POLICY":  cheWorkspaceSidecarImagePullPolicy,
```
Comment (davidfestal) [inline pkg/deploy/che_configmap.go]: > CheWorkspaceSidecarImagePullPolicy
> CheDockerAlwaysPullImage

it seems you don't do anything with these values here since you already added them in the main config map building code.
Either we given them a specific value for airgap, or we can just remove them completely and the default will be used by the Che server from the `che.properties`.

File: pkg/deploy/che_configmap.go (hunk 2 above)
Comment (nickboldt) [reply pkg/deploy/che_configmap.go]: yes, otherwise it'll try to pull from dockerhub/quay/rhoi repos.

File: pkg/deploy/che_configmap.go (hunk 3 above)
Comment (nickboldt) [reply pkg/deploy/che_configmap.go]: yeah we might be OK here since the assumption is that the newer images will also be inside the airgap environment. 

As long as the image r.r.io/crw/server:2.0 -> myquay.bigcorp.com/crw-stuff/crw-server:2.0 and newer ones are managed under the same tag... then no policy change should be needed.

I mean pullpolicy of Always pull might annoy teh devops team at bigcorp as it's extra network traffic on every startup but that'd be up to them to decide when installing Che/CRW in teh first place... right?

Comment (davidfestal) [reply pkg/deploy/che_configmap.go]: yeah. So let not override those at all for now

**Comment by davidfestal** [comment]
I would say LGTM unless you @l0rd and @nickboldt still want to approve before.

File: pkg/deploy/extra_images.go [hunk 5]
```diff
@@ -0,0 +1,8 @@
+// This file is generated, and contains the latest versions of certain properties from che.properties
+package deploy
+
+const (
+	cheWorkspacePluginBrokerInitImage    = "eclipse/che-init-plugin-broker:v0.21"
```
Comment (nickboldt) [inline pkg/deploy/extra_images.go]: I would prefer if these contained the registry in the URL (docker.io) as a reminder of things we need to migrate to quay.

Also don't love that jwtproxy does SHA releases, not numbered ones (7.3.0). But those are out of scope for this issue. :)

File: pkg/deploy/che_configmap.go [hunk 6]
```diff
@@ -67,7 +67,9 @@ type CheConfigMap struct {
 	WebSocketEndpointMinor               string `json:"CHE_WEBSOCKET_ENDPOINT__MINOR"`
 	CheWorkspacePluginBrokerInitImage    string `json:"CHE_WORKSPACE_PLUGIN__BROKER_INIT_IMAGE,omitempty"`
 	CheWorkspacePluginBrokerUnifiedImage string `json:"CHE_WORKSPACE_PLUGIN__BROKER_UNIFIED_IMAGE,omitempty"`
-	CheServerSecureExposesJwtProxyImage  string `json:"CHE_SERVER_SECURE__EXPOSER_JWTPROXY_IMAGE,omitempty"`
+	CheServerSecureExposerJwtProxyImage  string `json:"CHE_SERVER_SECURE__EXPOSER_JWTPROXY_IMAGE,omitempty"`
+	CheWorkspaceSidecarImagePullPolicy   string `json:"CHE_WORKSPACE_SIDECAR_IMAGE__PULL__POLICY,omitempty"`
+	CheDockerAlwaysPullImage             string `json:"CHE_DOCKER_ALWAYS__PULL__IMAGE,omitempty"`
```
[the remark this answers is not in the document]
Comment (sleshchenko) [reply pkg/deploy/che_configmap.go]: And probably override for [plugin_broker.pull_policy](https://github.com/eclipse/che/blob/36fe587e28fba1e37d9f6b38da673cf917a5132f/assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties#L633) should be added

[the remark this answers is not in the document]
Comment (sleshchenko) [reply pkg/deploy/che_configmap.go]: maybe imagePullPolicy for jwtproxy should be overridden as well https://github.com/eclipse/che/blob/154d668505cc86763f16ad8868e0a1fcf4bccc28/infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/secure/jwtproxy/JwtProxyProvisioner.java#L287

Some of my messages may be out of the topic since I'm not familiar with airgap topic but I have seen familiar words and thought maybe by shared knowledge can be useful here =)

[the remark this answers is not in the document]
Comment (davidfestal) [reply pkg/deploy/che_configmap.go]: After speaking with @nickboldt we had concluded that the airgap scenario doesn't need any overriding of the default pull policies. Even a `Always` pull policy would work in airgap mode, since we still point to a valid container registry.

So I'd better leave everything non-airgap-related out of the scope of this issue / PR.

Do you agree @sleshchenko ?

[the remark this answers is not in the document]
Comment (davidfestal) [reply pkg/deploy/che_configmap.go]: BTW both properties:
```
CHE_WORKSPACE_SIDECAR_IMAGE__PULL__POLICY
CHE_DOCKER_ALWAYS__PULL__IMAGE
```
Have already removed by Tom.

**Comment by l0rd** [comment]
Good job @tomgeorge. Would you mind updating the description of this PR? I think that you are not using the `airgapMode` attribute anymore. For future references it's better if the description corresponds to what has been merged.

**Comment by tomgeorge** [comment]
@l0rd I updated the description.  Thanks!
